go/tools/bazel/runfiles/resolver.go

```diff
@@ -0,0 +1,61 @@
+// Copyright 2018 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package runfiles
+
+import (
+	"errors"
+	"os"
+)
+
+const (
+	RUNFILES_MANIFEST_FILE = "RUNFILES_MANIFEST_FILE"
+	RUNFILES_DIR           = "RUNFILES_DIR"
+)
+
+// These are the possible Runfiles-related errors.
+var (
+	ErrNoRunfilesEnv   = errors.New("runfiles environment missing")
+	ErrManifestInvalid = errors.New("runfiles manifest syntax error")
+)
+
+// Resolver is an interface for a resolver that can take a runfiles path and
+// resolve it to a path on disk.
+type Resolver interface {
+	Resolve(string) (string, error)
+}
+
+// NewResolver creates a new runfiles resolver. The type of resolver and its
+// parameters are derived from the environment.
+func NewResolver() (Resolver, error) {
+	return newResolverWithGetenv(os.Getenv)
+}
+
+func newResolverWithGetenv(getenv func(string) string) (Resolver, error) {
```

Comment:
I'd rather not inject a test replacement for `getenv`. It'd be way too easy to accidentally call `os.Getenv` explicitly in a future change.

I think it's better for the tests to explicitly set the environment variables. To make sure they get cleaned up at the end of each test case, define a helper function like this:

```
defer setenvForTest(key, value string) (cleanup func()) {
  if old, ok := os.LookupEnv(key); ok {
    cleanup = func() { os.Setenv(key, old) }
  } else {
    cleanup = func() { os.Unsetenv(key) }
  }
  os.Setenv(key, value)
  return cleanup
}
```

Then call like this:

```
cleanupEnv := setenvForTest(RUNFILES_MANIFEST_FILE, manifestfn)
defer cleanupEnv()
```

Reply:
Works for me, Done. I am paranoid relying on serial execution of tests, but it seems tests do not run in parallel by default, so it's a non-issue.